herddb-core/src/main/java/herddb/core/TableManager.java

```diff
@@ -1123,18 +1137,25 @@ private static LockHandle lockForRead(Bytes key, Transaction transaction, String
         }
         List<UniqueIndexLockReference> uniqueIndexes = null;
         Map<String, AbstractIndexManager> indexes = tableSpaceManager.getIndexesOnTable(table.name);
-        if (indexes != null) {
+        if (indexes != null || table.foreignKeys != null) {
             try {
                 DataAccessor values = new Record(key, Bytes.from_array(value)).getDataAccessor(table);
-                for (AbstractIndexManager index : indexes.values()) {
-                    if (index.isUnique()) {
-                        Bytes indexKey = RecordSerializer.serializeIndexKey(values, index.getIndex(), index.getColumnNames());
-                        if (uniqueIndexes == null) {
-                            uniqueIndexes = new ArrayList<>(1);
+                if (table.foreignKeys != null) {
+                    for (ForeignKeyDef fk : table.foreignKeys) {
+                        checkForeignKeyConstraintsAsChildTable(fk, values, context, transaction);
```

Comment:
You are right.
Support for "DEFERRED" will come in a follow up work (if needed)

btw here we are going to hold a read lock inside the transaction (if a transaction is active) and this will guarantee that the table is in a consistent state during the whole time of the transaction